add-azure-versioning-guide-draft (#36740)
Azure/azure-rest-api-specs: documentation/versioning.md

        └── <RPNamespace>/		ß NOTE: Control-plane only (not data-plane) 
            └── <ServiceName1>/	ß CX-facing service name; each version gets Docs & SDK package
                ├── tspconfig.yaml	ß TypeSpec files go here (examples below)
                ├── main.tsp
                ├── models.tsp
                ├── readme.md		ß autorest readme with YAML blocks
                └── examples/		ß TypeSpec example folder
                    └── <api-version>/	ß One folder per service version described in TypeSpec
                        └── <example .json files> 
                └── preview/ and stable/
                    └── <api-version>/	ß One folder per service version described in Swagger. These folders are created and populated by compiling the TypeSpec folder for the service.
                        ├── <swagger .json files>
                        └── examples/	ß Swagger example folder
                            └── <example .json files>
            └── <ServiceName2>/	// CX-facing service name; contents identical to above structure
    └── data-plane/
        └── <ServiceName3>/	// CX-facing service name; contents identical to above structure
```

> In this structure, Azure services are presented as an RPNamespace that is used to manage services (both with customer-facing names)., each service has versions (some preview and some stable), and each version has its own API contract, documentation, and (beta/GA) SDK package(s).   

Reviewer: Service versions are preview or GA, SDK packages are beta or stable.

Author: Thanks for the suggestion, I have changed everywhere for stable to GA when it's about service version.

To enforce these rules, Here're some tweaks to the repo structure:

1.	In the resource-manager folder, if it's RPaaS service, there must be just 1 readme.md file, and 1+ service name folders under resource-manager/RPnamespace folder, but the resource types within each services folder must not duplicate except for operations, locations, checkNameAvailability etc. Which is required by RPSaaS for HTTP request schema validation.
2.	In the data-plane folder there must be no files and at least 1 service name folder.
3.	The ServiceName folder must have only readme.*, tspconfig.yaml and *.tsp files in it. 

>     - In the readme.md file, the latest package tag value (which represents a service version) must refer to swagger files that are all in the exact same folder. Also, if the tag value is suffixed with “-preview” then all the swagger paths must refer to files in the preview subfolder; else all swagger paths must refer to files in the stable subfolder.

Author: OpenAPI should be fine?

>     - NOTE: If 2+ services want to “share” some TypeSpec/swagger files, these files must be explicitly copied into a consuming service’s folder on its desired timeline. We do not allow multiple services to reference a single copy of a service contract because it is too dangerous to allow one service to change a shared contract, forcing other services to pick up this shared contract instantly, especially if the shared contract introduces breaking changes. Each service must be in control of its own destiny.

Author: I have changed everywhere from swagger into OpenAPI.

6.	The preview and stable folders must only have subfolders whose name match this format: YYYY-MM-DD-preview in the preview folder and YYYY-MM-DD in the stable folder. All the files in the YYYY-MM-DD(-preview) folders must be json (OpenAPI) files.
7.	Each YYYY-MM-DD(-preview) folder must have an examples folder containing only example *.json files for this api-version.
8.	It is illegal for a preview and GA version to share the same date. For example, if there is a 2024-04-01-preview, then there can never be a 2024-04-01 GA.
9. Each new control plane api-version should be unique across all the control plane service folders for an RP Namespace. 
10.	Each service name folder must duplicate the special "list operations" API. 
    - While this RP (not service) operation requires an api-version query parameter, the response data is not specific to this api-version; the response is api-version neutral. In addition, the response is always the same regardless of tenant/subscription. Whenever any other service in the RP gets a new api-version, the TypeSpec/OpenAPI for the operations service must also be updated to match this new api-version.

> ## POSITIVE CONSEQUENCES OF THE AZURE SERVICE DEFINITION

Author: ```suggestion
## Positive Consequences of the Azure Service Definition
```

                ├── readme.md		ß autorest readme with YAML blocks
                └── examples/		ß TypeSpec example folder
                    └── <api-version>/	ß One folder per service version described in TypeSpec
                        └── <example .json files> 
                └── preview/ and stable/
                    └── <api-version>/	ß One folder per service version described in Swagger
                        ├── <swagger .json files>
                        └── examples/	ß Swagger example folder
                            └── <example .json files>
            └── <ServiceName2>/	// CX-facing service name; contents identical to above structure
    └── data-plane/
        └── <ServiceName3>/	// CX-facing service name; contents identical to above structure
```
We don’t want internal teams or customers to think about the orgName, we want customers to think that Azure has RPNamespaces that are used to manage services (both with customer-facing names) where each service has versions (some preview and some stable), & each version has its own API contract, documentation, and (beta/GA) SDK package(s).   

To enforce these rules, Here're some tweaks to the repo structure:

> 1.	Force the folder immediately under specification to be the RPNamespace name (replacing orgName). EngSys should validate that this folder name matches the RPnamespace name currently under the resource-manager folder and that there are no files in this folder and that the only subfolders are resource-manager and data-plane. 

Author: @JeffreyRichter I wonder if we can delete this entry? As it would be misleading to services team and this is not what we currently do.

    - Ideally, we’d get rid of the resource-manager folder’s RPNamespace subfolder as this extra level in the hierarchy serves no purpose and is just confusing, but this can be done in a later phase.

> 2.	In the resource-manager folder, if it's RPaaS service, there must be just 1 readme.md file, and 1+ service name folders under resource-manager/RPnamespace folder, but the resource types within each services folder must not duplicate except for operations, locations, checkNameAvailability etc. Which is required by RPSaaS for HTTP request schema validation.

Author: This is also modified based on our latest conversation with RPaaS team.

5.	The ServiceName folder must have examples, preview and stable subfolders. Additional subfolders are allowed to help organize *.tsp files. The rules should ensure that only the tspconfig.yaml  and *.tsp files are in the ServiceName folder.  
6.	The examples folder should only have a subfolder for each api-version and each api-version folder can contain only example *.json files for this api-version.
7.	The preview and stable folders must only have subfolders whose name match this format: YYYY-MM-DD-preview in the preview folder and YYYY-MM-DD in the stable folder. All the files in the YYYY-MM-DD(-preview) folders must be json (swagger) files.
8.	Each YYYY-MM-DD(-preview) folder must have an examples folder containing only example *.json files for this api-version.
9.	It is illegal for a preview & stable version to share the same date. For example, if there is a 2024-04-01-preview, then there can never be a 2024-04-01 stable. It would be great for the repo tooling to detect this and prevent merging of the PR. 

> 10. It is not recommended if different service name folder share the same apiVersion when service adds a new version. 

Author: @JeffreyRichter I added this entry here as we talked last week.

> 11.	Each service name folder must duplicate the special “Operations operation” (see RPC docs, real-life docs and folder example). 

Author: I modify this part based on our latest practice.

Reviewer 2: Here I think we mean the 'list operations'  api that is required by ARM?

Author: Yes, in Jeffrey's docx, it's still the old version where there's to have a presudo operations service folder, but due to our latest practice and confirm with ARM team, we recommend service team to duplicate the operations list api in each service folder.

Author: Do we need to maintain two copies, I thought we are changing the source of truth to this one? If necessary I will make the changes back to his doc.

Reviewer 3: I think just proceeding with the Wiki is OK.
The only downside I see is that the link to the word doc has been given out a LOT over the past few years and will probably continue to be forwarded to people and this will not be out of date.

@qiaozha: Could you update my Word doc at the very top in a big font with yellow highlight that the doc has been replaced with the Wiki and put the link to the Wiki in the Word doc. Also add that the doc will not be kept up to date - the Wiki is the new source of truth.

Thank you.

Author: here's the wiki link https://github.com/Azure/azure-rest-api-specs/wiki/Azure-Service-Versioning-Guideline and the docx is also updated. Thanks!

    - While this RP (not service) operation requires an api-version query parameter, the response data is not specific to this api-version; the response is api-version neutral. In addition, the response is always the same regardless of tenant/subscription. Whenever any other service in the RP gets a new api-version, the TypeSpec/Swagger for the operations service must also be updated to match this new api-version.

## POSITIVE CONSEQUENCES OF THE AZURE SERVICE DEFINITION

From the above definition, many processes and assets naturally flow, providing both Azure engineering teams and customers with great experiences:
- Each service version is documented separately allowing a customer to reference the specific version of the service they are targeting (example).
- Each service version gets its own SDK package version allowing a customer to import the SDK package version corresponding to the service version they wish to target.

>     - Also, stable versions of a service get a GA SDK package and preview versions of a service get a beta SDK package. Each SDK package version must clearly document which service version it targets .

Reviewer 2: ```suggestion
    - Stable versions of a service get a GA SDK package and preview versions of a service get a beta SDK package. Each SDK package version must clearly document which service version it targets .
```

>     - When any preview or stable version of a service is retired, the corresponding SDK package version is deprecated (without affecting any other SDK packages). Note that Azure retires preview service versions regularly; stable service versions are retired when breaking changes are introduced or on other rare occasions. See Azure SDK lifecycle and support policy.

Reviewer 2: Link to Azure SDK lifecycle and support policy?

    - When a service is retired in its entirety, all SDK packages for this service are deprecated (without affecting any other services’ SDK packages).

>     - NOTE: A service version introducing a breaking change will likely break that service’s SDK package in the next version causing the new SDK package version to increment its major number as per semantic versioning. This also causes old preview and stable service versions and their corresponding SDK package versions to be retired/deprecated.

Reviewer 2: ```suggestion
    - A service version introducing a breaking change will likely break that service’s SDK package in the next version causing the new SDK package version to increment its major number as per semantic versioning. This also causes old preview and stable service versions and their corresponding SDK package versions to be retired/deprecated.
```

> - The above service & SDK package versioning rules simplify the usage and on-going maintenance of downstream consumers such as the Azure Portal, Azure CLI/PowerShell, Azure Developer CLI, ARM templates, Bicep, 3rd party libraries that wrap our Azure REST APIs and SDK packages (such as Terraform/Pulumi), etc.  

Reviewer 2: ```suggestion
- The above service and SDK package versioning rules simplify the usage and ongoing maintenance of downstream consumers such as the Azure Portal, Azure CLI/PowerShell, Azure Developer CLI, ARM templates, Bicep, 3rd party libraries that wrap our Azure REST APIs and SDK packages (such as Terraform/Pulumi), etc.  
```

In general, we should remove usages of ampersand in favor of 'and'

The following diagram summarizes how Azure internal teams and customers should think about Azure service abstractions and their relationship to each other. 
- The RP namespace identifies the Resource Provider that customers must use to manage some Azure resource(s). The namespace itself has no code and therefore never versions. 

> - Within an RP namespace is 1+ control plane services and 0+ data-plane which do version uniformly over time. 

Reviewer 2: ```suggestion
- Within an RP namespace is 1+ control plane services and 0+ data-plane service(s) which version uniformly over time. 
```

The last service to split-off from the monolithic SDK package may either get its own/new SDK package and name (as described in the previous paragraph) or it may stay in a new version of the original SDK package (same name as before). But, in this new package version, everything related to the already split-out services no longer exists. This latter scenario reduces some customer pain, but the name of the package may not accurately reflect the name of the remaining service. We’ll have to decide on the package name on a case-by-case basis.  

The Azure Breaking Change and Azure SDK Architecture Boards are aware of this scenario and have agreed to approve these 1-time breaks related to splitting a monolithic SDK package into multiple service-specific packages so that engineering and customers can get on a long-term sustainable future. The Azure API Stewardship and ARM review Boards are also aware of this.  

> As of 8/19/25, several Azure service teams have already decided to split their monolithic SDK package into service specific SDK packages: Compute, Monitor, Container Registry, Observability, AI Language, AI Speech, and more.  

Reviewer 2: I also think we shouldn't have a date here, let's just list several important services that have completed this work
